src/test/resources/com/puppycrawl/tools/checkstyle/grammar/antlr4/InputAntlr4AstRegressionPatternMatchingInSwitch.txt

```diff
@@ -0,0 +1,1012 @@
+        |   `--SLIST -> { [42:29]
+        |       |--LITERAL_SWITCH -> switch [43:8]
+        |       |   |--LPAREN -> ( [43:15]
+        |       |   |--EXPR -> EXPR [43:16]
+        |       |   |   `--IDENT -> o [43:16]
+        |       |   |--RPAREN -> ) [43:17]
+        |       |   |--LCURLY -> { [43:19]
+        |       |   |--CASE_GROUP -> CASE_GROUP [44:12]
+        |       |   |   |--LITERAL_CASE -> case [44:12]
+        |       |   |   |   |--LAND -> && [44:26]
+        |       |   |   |   |   |--PATTERN_VARIABLE_DEF -> PATTERN_VARIABLE_DEF [44:17]
+        |       |   |   |   |   |   |--MODIFIERS -> MODIFIERS [44:17]
+        |       |   |   |   |   |   |--TYPE -> TYPE [44:17]
+        |       |   |   |   |   |   |   `--IDENT -> String [44:17]
+        |       |   |   |   |   |   `--IDENT -> s [44:24]
+        |       |   |   |   |   `--GT -> > [44:40]
+        |       |   |   |   |       |--METHOD_CALL -> ( [44:37]
+        |       |   |   |   |       |   |--DOT -> . [44:30]
+        |       |   |   |   |       |   |   |--IDENT -> s [44:29]
+        |       |   |   |   |       |   |   `--IDENT -> length [44:31]
+        |       |   |   |   |       |   |--ELIST -> ELIST [44:38]
+        |       |   |   |   |       |   `--RPAREN -> ) [44:38]
+        |       |   |   |   |       `--NUM_INT -> 4 [44:42]
+        |       |   |   |   `--COLON -> : [44:43]
+        |       |   |   `--SLIST -> SLIST [45:34]
+        |       |   |       |--EXPR -> EXPR [45:34]
+        |       |   |       |   `--METHOD_CALL -> ( [45:34]
+        |       |   |       |       |--DOT -> . [45:26]
+        |       |   |       |       |   |--DOT -> . [45:22]
+        |       |   |       |       |   |   |--IDENT -> System [45:16]
+        |       |   |       |       |   |   `--IDENT -> out [45:23]
+        |       |   |       |       |   `--IDENT -> println [45:27]
+        |       |   |       |       |--ELIST -> ELIST [45:35]
+        |       |   |       |       |   `--EXPR -> EXPR [45:35]
+        |       |   |       |       |       `--IDENT -> s [45:35]
+        |       |   |       |       `--RPAREN -> ) [45:36]
+        |       |   |       |--SEMI -> ; [45:37]
+        |       |   |       `--LITERAL_BREAK -> break [46:16]
+        |       |   |           `--SEMI -> ; [46:21]
+        |       |   |--CASE_GROUP -> CASE_GROUP [47:12]
+        |       |   |   |--LITERAL_CASE -> case [47:12]
+        |       |   |   |   |--LAND -> && [47:27]
+        |       |   |   |   |   |--PATTERN_VARIABLE_DEF -> PATTERN_VARIABLE_DEF [47:17]
+        |       |   |   |   |   |   |--MODIFIERS -> MODIFIERS [47:17]
+        |       |   |   |   |   |   |--TYPE -> TYPE [47:17]
+        |       |   |   |   |   |   |   `--IDENT -> Integer [47:17]
+        |       |   |   |   |   |   `--IDENT -> i [47:25]
+        |       |   |   |   |   `--GT -> > [47:32]
+        |       |   |   |   |       |--IDENT -> i [47:30]
+        |       |   |   |   |       `--NUM_INT -> 40 [47:34]
```

Comment:
Corresponding code:

```java

            case Integer i && i > 40:
                System.out.println("Integer");

```

Should we add a new token, `PATTERN_DEF` ?

Example:

```
|-PATTERN_DEF -> [47:27]        
|   |--LAND -> && [47:27]
|   |   |--PATTERN_VARIABLE_DEF -> PATTERN_VARIABLE_DEF [47:17]
|   |   |   |--MODIFIERS -> MODIFIERS [47:17]
|   |   |   |--TYPE -> TYPE [47:17]
|   |   |   |   `--IDENT -> Integer [47:17]
|   |   |   `--IDENT -> i [47:25]
|   |   `--GT -> > [47:32]
|   |       |--IDENT -> i [47:30]
|   |       `--NUM_INT -> 40 [47:34]

```

This would apply only to applications of the `pattern` production rule, and would have no effect on existing `o instanceof String s` definitions (causing no changes to existing AST structure). As pattern definitions become more prevalent in Java grammar, it may prove helpful to have a dedicated `PATTERN_DEF` node.

Reply:
@romani @pbludov @rnveach @strkkk please share your thoughts.

Reply:
Am I got it right,

```java
            case Integer i && i < 40:
                System.out.println("Small integer");
                break;
```
will be
```
|-PATTERN_DEF -> [47:27]        
|   |--LAND -> && [47:27]
|   |   |--PATTERN_VARIABLE_DEF -> PATTERN_VARIABLE_DEF [47:17]
|   |   |   |--MODIFIERS -> MODIFIERS [47:17]
|   |   |   |--TYPE -> TYPE [47:17]
|   |   |   |   `--IDENT -> Integer [47:17]
|   |   |   `--IDENT -> i [47:25]
|   |   `--LT -> > [47:32]
|   |       |--IDENT -> i [47:30]
|   |       `--NUM_INT -> 40 [47:34]
```
and

```java
            case Integer i:
                System.out.println("Any integer");
                break;
```
will be also a `PATTERN_DEF`: 
```
|-PATTERN_DEF -> [47:27]        
|   |--PATTERN_VARIABLE_DEF -> PATTERN_VARIABLE_DEF [47:17]
|   |   |--MODIFIERS -> MODIFIERS [47:17]
|   |   |--TYPE -> TYPE [47:17]
|   |   |   `--IDENT -> Integer [47:17]
|   |   `--IDENT -> i [47:25]
```
?

Reply:
> will be also a PATTERN_DEF:
> ?

We can do this.

Or, we can have a simple `PATTERN_VARIABLE_DEF` with no `PATTERN_DEF` node for switch, just like we do for instanceof. I am not sure what will prove useful from a static analysis viewpoint, because I can't really think of how any current checks would benefit from more easily accessing patterns in the AST (by adding a new token). I can work up some more examples if we want to consider adding a new token.

Reply:
I think having `PATTERN_DEF` over `LAND` makes sense. This will allow some checks to distinguish between `&&` for patterns and conditions.

Adding 'PATTERN_DEF` as the parent of 'PATTERN_VARIABLE_DEF' will be a breaking change for some third-party checks.

I like simple `PATTERN_VARIABLE_DEF` for switch.

Reply:
we have now PATTERN_VARIABLE_DEF
https://github.com/checkstyle/checkstyle/blob/272dbe394e1d581aec6743a7c7eba0aab0d61f9b/src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java#L5540-L5543

Reply:
why we think that it is not CASE specific syntax for pattern ?  Should we name token like CASE_PATTERN_DEF ? to keep it specific and let it always handle all switch-case specific nuances.

Reply:
One more thought: the `if` has no `PATTERN_DEF`:
```java
if (o instanceof Integer i && i < 40) { // LAND, LITERAL_INSTANCEOF, PATTERN_VARIABLE_DEF, LT
 ...
}

switch (o) {
  case Integer i && i < 40: // LAND, PATTERN_VARIABLE_DEF, LT ?
    break;
}
````
Should there be an extra node for `case`?

Reply:
> why we think that it is not CASE specific syntax for pattern

It is not. This syntax can be used anywhere enhanced instance of with pattern matching can be used now, as long as expression is parenthesized. 

<details>
 <summary>Example:</summary>

```
public class Test {
    void method1(Object o) { // should have violation
        if (o instanceof String s) {
            // ^ pattern variable introduced (simple type pattern)
        }
        
        if (o instanceof String s && s.length() > 6) {
            // ^ pattern variable introduced (simple type pattern),
            // then evaluated in boolean expression
            // 's.length() > 6' is usual boolean expression, not part of pattern
        }
        
        if (o instanceof (String s && s.length() > 4)) {
            // ^ pattern variable introduced in guarded pattern; this refines the pattern itself
            // 's.length() > 4` is part of the pattern
        }
        
        int x = o instanceof (String s && s.length() > 6) ? 4 : 5;
        int y = o instanceof String s && s.length() > 4 ? 2 : 3;
        int z = o instanceof String s ? 1 : 2;
        
        while (o instanceof (String s && s.length() > 6)) {
            
        }
        while (o instanceof String s && s.length() > 4) {
            
        }
        while (o instanceof String s) {
            
        }
        
        for (int i = 0; o instanceof Integer myInt && myInt > 5;) {
            
        }
        for (int i = 0; o instanceof (Integer myInt && myInt > 5);) {
            
        }
        for (int i = 0; o instanceof Integer myInt; ) {
            
        }
    }
}

```
</details>

> to keep it specific...

To me, specificity is more important for the difference between a simple pattern variable introduction, like:

```
        if (o instanceof String s) {
            // ^ pattern variable introduced (simple type pattern)
        }

```

And 

```
        if (o instanceof (String s && s.length() > 4)) {
            // ^ pattern variable introduced in guarded pattern; this refines the pattern itself
            // 's.length() > 4` is part of the pattern
        }

```

In simple pattern variable introduction, whatever surrounding context we need can already be gathered by a check (and probably already is). This does not require (or benefit from) a new parent `PATTERN_DEF` (and would cause breaking changes to AST.

In guarded and parenthesized patterns, we should be able to know exactly all that this pattern encompasses easily. Why? Because pattern definitions are going to get much more complicated soon due to pattern matching for records, in [JEP405](https://openjdk.java.net/jeps/406):

```
static void printXCoordOfUpperLeftPointWithPatterns(Rectangle r) {
    if (r instanceof Rectangle(ColoredPoint(Point(var x, var y), var c), var lr)) {
        System.out.println("Upper-left corner: " + x);
    }
}

static void printSumOfFirstTwoXCoords(Object o) {
    if (o instanceof Point[] { Point(var x1, var y1), Point(var x2, var y2), ... }) {
        System.out.println(x1 + x2);
    }
}

```

In this case, we can easily have entire AST for `Rectangle(ColoredPoint(Point(var x, var y), var c), var lr)` in one go, and have a parent to look for when analyzing scopes of pattern variables, etc.

Reply:
> One more thought: the if has no PATTERN_DEF

Without parenthesized pattern, it would not. It's not really necessary, since `&& i < 40` is not part of the pattern. If a parenthesized pattern was used (`o instanceof (Integer i && i < 40)`), then it would now be a `PATTERN_DEF` by my proposal, because `&& i < 40` is now part of the pattern.

Reply:
How about this: I will make a few small/simple input files, with examples from above, for evaluation, and add to this PR, using proposed implementation. This will make it easier to visualize exact AST for certain syntax. Thumbs up on this for agreement.

Reply:
OK, implementation and tests updated. I am marking this ready for review.

Reply:
OK, implementation and tests updated. I am marking this ready for review.